### ISSUE TITLE
Sieve of Eratosthenes for a stream of all prime numbers

### ISSUE THREAD

**Issue by Taneb** [agda/agda-stdlib#1859]
(this is very much high-hanging fruit, or at least I've tried to implement it a few times and gotten stuck)

It would be really useful to have a way of generating prime numbers in stdlib.

Here's a sketch of an API:

```agda
primes : Stream ℕ
primes-all-prime : All Prime primes
primes-ordered : Linked _<_ primes -- need to define Linked for Stream
primes-complete : ∀ {n} → Prime n → n ∈ primes
```

It ought to be possible to adapt Melissa O'Neill's paper "The Genuine Sieve of Eratosthenes" for this, but that would ideally use a priority queue (see #1542), and when I've tried in the past I've run into termination issues that I've not been able to resolve.

**Comment by MatthewDaggitt** [comment]
I think any implementation even without a priority queue would be a good first step!

**Comment by Taneb** [comment]
Yeah, I agree. It'll work just as well with the already-defined AVL trees, albeit slightly slower.

**Comment by gallais** [comment]
Yves Bertot has a paper doing just that in Coq. I have parts of it in Agda here but never found
the time to complete the proof: https://github.com/gallais/potpourri/tree/main/agda/papers/sieve04

**Comment by Taneb** [comment]
Oh, that's very interesting! I'll make sure I give that a read
